test/extended/images/helper.go

```diff
@@ -423,3 +588,127 @@ func IsBlobStoredInRegistry(
 	}
 	return present, presentInRepository, err
 }
+
+// RunHardPrune executes inside a docker-registry pod and runs a garbage collector. The docker-registry is
+// assumed to be in a read-only mode and using filesystem as a storage driver. It returns lists of deleted
+// files.
+func RunHardPrune(oc *exutil.CLI) (*RegistryStorageFiles, error) {
+	podList, err := oc.AsAdmin().KubeClient().Core().Pods("default").List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{"deploymentconfig": "docker-registry"}).String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(podList.Items) == 0 {
+		return nil, fmt.Errorf("failed to find any docker-registry pod")
+	}
+
+	defer func(ns string) { oc.SetNamespace(ns) }(oc.Namespace())
+	output, err := oc.AsAdmin().SetNamespace(metav1.NamespaceDefault).Run("env").Args("--list", "dc/docker-registry").Output()
+	if err != nil {
+		return nil, err
+	}
+
+	podName := podList.Items[0].Name
+	deleted := &RegistryStorageFiles{
+		Repos:         []string{},
+		ManifestLinks: make(RepoLinks),
+		LayerLinks:    make(RepoLinks),
+		Blobs:         []string{},
+	}
+
+	err = wait.ExponentialBackoff(retry.DefaultBackoff, func() (bool, error) {
+		out, err := oc.SetNamespace(metav1.NamespaceDefault).AsAdmin().
+			Run("exec").Args("--stdin", podName, "--", "/bin/sh", "-s").
+			InputString(fmt.Sprintf(registryGCLauncherScript, dockerRegistryBinary)).Output()
+		if exitError, ok := err.(*exutil.ExitError); ok && strings.Contains(exitError.StdErr, "unable to upgrade connection") {
```

Comment:
@stevekuznetsov have you seen this before?

```
Jun 16 12:27:17.529: INFO: Running 'oc exec --config=/tmp/openshift/core/openshift.local.config/master/admin.kubeconfig --namespace=default --stdin docker-registry-1-z3hj1 -- /bin/sh -s'
Jun 16 12:27:17.771: INFO: Error running &{/data/src/github.com/openshift/origin/_output/local/bin/linux/amd64/oc [oc exec --config=/tmp/openshift/core/openshift.local.config/master/admin.kubeconfig --namespace=default --stdin docker-registry-1-z3hj1 -- /bin/sh -s] []   error: unable to upgrade connection: container not found ("registry")
 error: unable to upgrade connection: container not found ("registry")
 [] <nil> 0xc420d582d0 exit status 1 <nil> <nil> true [0xc420030000 0xc420030040 0xc420030040] [0xc420030000 0xc420030040] [0xc420030020 0xc420030038] [0xde31c0 0xde32c0] 0xc421682240 <nil>}:
error: unable to upgrade connection: container not found ("registry")
failed to execute into registry pod docker-registry-1-z3hj1: exit status 1
Jun 16 12:27:17.822: INFO: Running 'oc exec --config=/tmp/openshift/core/openshift.local.config/master/admin.kubeconfig --namespace=default --stdin docker-registry-1-z3hj1 -- /bin/sh -s'
Jun 16 12:27:18.065: INFO: Error running &{/data/src/github.com/openshift/origin/_output/local/bin/linux/amd64/oc [oc exec --config=/tmp/openshift/core/openshift.local.config/master/admin.kubeconfig --namespace=default --stdin docker-registry-1-z3hj1 -- /bin/sh -s] []   error: unable to upgrade connection: container not found ("registry")
 error: unable to upgrade connection: container not found ("registry")
 [] <nil> 0xc4212150b0 exit status 1 <nil> <nil> true [0xc4202b76d8 0xc4202b7780 0xc4202b7780] [0xc4202b76d8 0xc4202b7780] [0xc4202b76f8 0xc4202b7760] [0xde31c0 0xde32c0] 0xc42168ef00 <nil>}:
error: unable to upgrade connection: container not found ("registry")
```

Do I use the `exec` wrong? The interesting part is that this fails roughly in one out of three runs - the first time this code is executed.
If this succeeds once, it will succeed next time as well.